File: frontend/src/types/metrics.ts
import type { EmployeeLevel } from "./employee";

// Shape follows PRD.md §24 (Dashboard API Response). The frontend renders these
// values but does not recompute leadership metric business logic.

export type DashboardScope =
  | { type: "org" }
  | { type: "manager"; id: string }
  | { type: "level"; id: EmployeeLevel };

export interface ScopeDescriptor {
  type: "org" | "manager" | "level";
  id: string | null;
  name: string;
}

export interface Coverage {
  eligible_employees: number;
  respondents: number;
  response_rate: number;
}

export interface Population {
  eligible_employees: number;
  respondents: number;
  non_respondents: number;
  active_ai_users: number;
}

export interface RateMetric {
  value: number;
  count: number;
  denominator: number;
}

export interface HeadlineMetrics {
  ai_adoption_rate: RateMetric;
  reports_more_output: RateMetric;
}

export interface DistributionRow {
  code: string;
  label: string;
  count: number;
  pct: number;
}

export interface QuestionDistribution {
  denominator: number;
  rows: DistributionRow[];
}

export interface ValueAreaRow {
  code: string;
  label: string;
  rank1: number;
  rank2: number;
  rank3: number;
  total: number;
  otherTexts: Record<string, number>;
}

export interface ValueAreaRanking {
  denominator: number;
  rows: ValueAreaRow[];
}

export interface RankedOptionRow {
  code: string;
  label: string;
  count: number;
  pct: number;
  otherTexts: Record<string, number>;
}

export interface OptionDistribution {
  denominator: number;
  rows: RankedOptionRow[];
}

export interface Q3Q5Criteria {
  weekly_time_saved: string;
  work_output_change: string;
  quality_change: string;
}

export interface Q3Q5Analysis {
  criteria: Q3Q5Criteria;
  matching_count: number;
  analysis_denominator: number;
  matching_rate: number;
}

export interface GroupRow {
  key: string;
  label: string;
  eligible_employees: number;
  respondents: number;
  adoption_rate: number | null;
  more_output_rate: number | null;
  frequent_rework_rate: number | null;
  top_barrier: { code: string; label: string } | null;
}

export interface GroupBreakdown {
  group_by: "level";
  rows: GroupRow[];
}

export interface DashboardMetricsResponse {
  scope: ScopeDescriptor;
  coverage: Coverage;
  population: Population;
  headline_metrics: HeadlineMetrics;
  usage_frequency: QuestionDistribution;
  workflow_value: ValueAreaRanking;
  weekly_time_saved: QuestionDistribution;
  work_output: QuestionDistribution;
  work_quality: QuestionDistribution;
  ai_rework_frequency: QuestionDistribution;
  q3_q5_analysis: Q3Q5Analysis;
  benefits: OptionDistribution;
  barriers: OptionDistribution;
  group_breakdown: GroupBreakdown;
}

File: backend/app/models/metrics.py
from typing import Literal

from pydantic import BaseModel, field_validator

from app.models.employee import EmployeeLevel
from app.models.survey_response import QualityChange, WorkOutputChange

ScopeType = Literal["org", "manager", "level"]
# Department lives on the Employee record as display context only — it is not
# a supported dashboard grouping dimension (see docs/ADR.md / CLAUDE.md
# architecture principle 7). Level is the only group_breakdown dimension.
GroupByField = Literal["level"]
AnalysisWeeklyTimeSaved = Literal["no_noticeable_time_saved", "less_than_1_hour", "1_5_hours", "more_than_5_hours"]


class ScopeDescriptor(BaseModel):
    type: ScopeType
    id: str | None
    name: str


class Coverage(BaseModel):
    eligible_employees: int
    respondents: int
    response_rate: float


class Population(BaseModel):
    eligible_employees: int
    respondents: int
    non_respondents: int
    active_ai_users: int


class RateMetric(BaseModel):
    value: float
    count: int
    denominator: int


class AverageMetric(BaseModel):
    value: float
    denominator: int


class HeadlineMetrics(BaseModel):
    ai_adoption_rate: RateMetric
    avg_weekly_hours_saved: AverageMetric
    estimated_weekly_hours_saved: float
    reports_more_output: RateMetric


class DistributionRow(BaseModel):
    code: str
    label: str
    count: int
    pct: int


class QuestionDistribution(BaseModel):
    denominator: int
    rows: list[DistributionRow]

    @property
    def rows_by_code(self) -> dict[str, DistributionRow]:
        return {row.code: row for row in self.rows}


class ValueAreaRow(BaseModel):
    code: str
    label: str
    rank1: int
    rank2: int
    rank3: int
    total: int
    otherTexts: dict[str, int]


class ValueAreaRanking(BaseModel):
    denominator: int
    rows: list[ValueAreaRow]

    @property
    def rows_by_code(self) -> dict[str, ValueAreaRow]:
        return {row.code: row for row in self.rows}


class RankedOptionRow(BaseModel):
    code: str
    label: str
    count: int
    pct: int
    otherTexts: dict[str, int]


class OptionDistribution(BaseModel):
    denominator: int
    rows: list[RankedOptionRow]

    @property
    def rows_by_code(self) -> dict[str, RankedOptionRow]:
        return {row.code: row for row in self.rows}


class Q3Q5Criteria(BaseModel):
    weekly_time_saved: AnalysisWeeklyTimeSaved
    work_output_change: WorkOutputChange
    quality_change: QualityChange

    @field_validator("weekly_time_saved")
    @classmethod
    def reject_not_sure(cls, value: AnalysisWeeklyTimeSaved) -> AnalysisWeeklyTimeSaved:
        if value == "not_sure":
            raise ValueError("not_sure is not valid for Q3-Q5 analysis criteria")
        return value


class Q3Q5Analysis(BaseModel):
    criteria: Q3Q5Criteria
    matching_count: int
    analysis_denominator: int
    matching_rate: float


class TopBarrier(BaseModel):
    code: str
    label: str


class GroupRow(BaseModel):
    key: str
    label: str
    eligible_employees: int
    respondents: int
    adoption_rate: int | None
    more_output_rate: int | None
    avg_hours_saved: float | None
    avg_hours_saved_denominator: int
    frequent_rework_rate: int | None
    top_barrier: TopBarrier | None


class GroupBreakdown(BaseModel):
    group_by: GroupByField
    rows: list[GroupRow]


class DashboardMetricsResponse(BaseModel):
    scope: ScopeDescriptor
    coverage: Coverage
    population: Population
    headline_metrics: HeadlineMetrics
    usage_frequency: QuestionDistribution
    workflow_value: ValueAreaRanking
    weekly_time_saved: QuestionDistribution
    work_output: QuestionDistribution
    work_quality: QuestionDistribution
    ai_rework_frequency: QuestionDistribution
    q3_q5_analysis: Q3Q5Analysis
    benefits: OptionDistribution
    barriers: OptionDistribution
    group_breakdown: GroupBreakdown
